src/SessionSetup/RetrieveInstancesWidget.cpp

```diff
@@ -4,27 +4,160 @@
+  state_machine_.setGlobalRestorePolicy(QStateMachine::RestoreProperties);
+
+  s_idle_.addTransition(this, &RetrieveInstancesWidget::LoadingStarted, &s_loading_);
+
+  s_loading_.assignProperty(ui_->projectComboBox, "enabled", false);
+  s_loading_.assignProperty(ui_->filterLineEdit, "enabled", false);
+  s_loading_.assignProperty(ui_->allCheckBox, "enabled", false);
+  s_loading_.assignProperty(ui_->reloadButton, "enabled", false);
+  s_loading_.addTransition(this, &RetrieveInstancesWidget::LoadingSuccessful, &s_idle_);
+  s_loading_.addTransition(this, &RetrieveInstancesWidget::LoadingFailed, &s_idle_);
+  s_loading_.addTransition(this, &RetrieveInstancesWidget::InitialLoadingFailed,
+                           &s_initial_loading_failed_);
+
+  s_initial_loading_failed_.assignProperty(ui_->projectComboBox, "enabled", false);
+  s_initial_loading_failed_.assignProperty(ui_->filterLineEdit, "enabled", false);
+  s_initial_loading_failed_.assignProperty(ui_->allCheckBox, "enabled", false);
+  s_initial_loading_failed_.addTransition(this, &RetrieveInstancesWidget::LoadingStarted,
+                                          &s_loading_);
+}
+
+std::optional<Project> RetrieveInstancesWidget::GetQSettingsProject() {
+  QSettings settings;
+
+  std::optional<Project> remembered_project;
+  if (settings.contains(kSelectedProjectIdKey)) {
+    remembered_project = Project{
+        settings.value(kSelectedProjectDisplayNameKey).toString() /* display_name */,
+        settings.value(kSelectedProjectIdKey).toString() /* id */
+    };
+  }
+  return remembered_project;
+}
+
+void RetrieveInstancesWidget::Start() {
+  state_machine_.setInitialState(&s_loading_);
+  state_machine_.start();
+
+  QSettings settings;
+  ui_->allCheckBox->setChecked(settings.value(kAllInstancesKey, false).toBool());
+
+  InitialLoad(GetQSettingsProject());
+}
+
+Client::InstanceListScope RetrieveInstancesWidget::GetInstanceListScope() {
+  return ui_->allCheckBox->isChecked() ? Client::InstanceListScope::kAllReservedInstances
+                                       : Client::InstanceListScope::kOnlyOwnInstances;
+}
+
+void RetrieveInstancesWidget::InitialLoad(const std::optional<Project>& remembered_project) {
+  emit LoadingStarted();
+  retreive_instances_->LoadProjectsAndInstances(remembered_project, GetInstanceListScope())
+      .Then(main_thread_executor_,
+            [widget = QPointer<RetrieveInstancesWidget>(this)](
+                ErrorMessageOr<RetrieveInstances::LoadProjectsAndInstancesResult> loading_result) {
+              if (widget == nullptr) return;
+
+              if (loading_result.has_error()) {
+                emit widget->InitialLoadingFailed();
+                widget->OnInstancesLoadingReturned(loading_result.error());
+                return;
+              }
+
+              CHECK(widget->ui_->projectComboBox->count() == 0);
+
+              RetrieveInstances::LoadProjectsAndInstancesResult& data{loading_result.value()};
```

Comment:
Please use a more descriptive variable name than "data"

Reply:
I called it `initial_load_result` now.